R/list-transpose.R

```diff
@@ -0,0 +1,98 @@
+#' Transpose a list
+#'
+#' @description
+#' `list_transpose()` turns a list-of-lists "inside-out"; it turns a pair of
+#' lists into a list of pairs, or a list of pairs into pair of lists. For
+#' example, if you had a list of length `n` where each component had values `a`
+#' and `b`, `list_transpose()` would make a list with elements `a` and
+#' `b` that contained lists of length n.
+#'
+#' It's called transpose because `x[["a"]][["b"]]` is equivalent to
+#' `transpose(x)[["b"]][["a"]]`, i.e. transposing a list flips the order of
+#' indices in a similar way to transposing a matrix.
+#'
+#' @param x A list of vectors to transpose.
+#' @param template A "template" that specifies the names of output list.
+#'   Usually taken from the name of the first element of `x`.
+#' @param simplify Should the result be simplified?
+#'   * `TRUE`: simplify or die trying.
+#'   * `NA`: simplify if possible.
+#'   * `FALSE`: never try to simplify, always leaving as a list.
+#'
+#'   Alternatively, a named list specifying the simplification by output column.
+#' @param ptype An optional vector prototype used to control the simplification.
+#'   Alternatively, a named list specifying the prototype by output column.
+#' @param default A default value to use if a value is absent of `NULL`.
+#'   Alternatively, a named list specifying the prototype by output column.
+#' @export
+#' @examples
+#' # list_transpose() is useful in conjunction with safely()
+#' x <- list("a", 1, 2)
+#' y <- x %>% map(safely(log))
+#' y %>% str()
+#' # Put all the errors and results together
+#' y %>% list_transpose() %>% str()
+#' # Supply a default result to further simplify
+#' y %>% list_transpose(default = list(result = NA)) %>% str()
+#'
+#' # list_transpose() will try to simplify by default:
+#' x <- list(list(a = 1, b = 2), list(a = 3, b = 4), list(a = 5, b = 6))
+#' x %>% list_transpose()
+#' # use simplify = FALSE to always return lists:
+#' x %>% list_transpose(simplify = FALSE) %>% str()
+#'
+#' # Provide explicit template if you know which elements you want to extract
+#' ll <- list(
+#'   list(x = 1, y = "one"),
+#'   list(z = "deux", x = 2)
+#' )
+#' ll %>% list_transpose()
+#' ll %>% list_transpose(template = c("x", "y", "z"))
+#'
+#' # And specify default if you want to simplify
+#' ll %>% list_transpose(c("x", "y", "z"), default = NA)
+list_transpose <- function(x, template = NULL, simplify = NA, ptype = NULL, default = NULL) {
+  vec_check_list(x)
+  if (length(x) == 0) {
+    return(list())
+  }
+
+  template <- template %||%
+    names(x[[1]]) %||%
+    cli::cli_abort("First element of {.arg x} is unnamed, please supply {.arg template}.")
```

Comment:
```suggestion
    cli::cli_abort("Must supply either {.arg template} or a named {.arg x}.")
```

Reply:
I don't understand why these fail:

```r
list_transpose(
  list(
    list(1, 2, 3),
    list(4, 5, 6),
    list(7, 8, 9)
  )
)
#> Error in `list_transpose()`:
#> ! First element of `x` is unnamed, please supply `template`.

list_transpose(
  list(
    a = list(1, 2, 3),
    b = list(4, 5, 6),
    c = list(7, 8, 9)
  )
)
#> Error in `list_transpose()`:
#> ! First element of `x` is unnamed, please supply `template`.
```

Reply:
`list_transpose()` no longer round-trips:

```r
x <- list(list(a = 1, b = 2), list(a = 3, b = 4), list(a = 5, b = 6))
x %>% list_transpose()
#> $a
#> [1] 1 3 5
#>
#> $b
#> [1] 2 4 6

x %>% list_transpose() |> list_transpose()
#> Error in `list_transpose()`:
#> ! First element of `x` is unnamed, please supply `template`.
```

Reply:
I made it stricter because naming seems like a good principle, but it does seem undesirable to break the symmetry, so I'll bring that back.